### PR TITLE
[stable/redis-ha] Fix split brain issue

### LINKED ISSUE #9488: [stable/redis-ha] server-0 failure can result in split-brain
**Is this a request for help?**: No

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Version of Helm and Kubernetes**:

Helm version:
```
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
```

Kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-24T06:54:59Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.2-gke.18", GitCommit:"5796233393d7bc034428de15191ad3d2eaff95fb", GitTreeState:"clean", BuildDate:"2018-11-08T20:49:08Z", GoVersion:"go1.10.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

**Which chart**:
redis-ha

**What happened**:
A single node failure (out of 3) resulted in a redis outage due to bad logic in redis-ha chart.

1) Server 0 crashed / had a network partition with the other nodes
2) Server 0 rebooted
3) Server 0 could not reach the other nodes while it was starting up so it decided it was the master (https://github.com/helm/charts/blob/master/stable/redis-ha/templates/redis-ha-configmap.yaml#L57)
4) Server 0 thinks it's master indefinitely and is waiting for other nodes to connect to it; it cannot learn about the other nodes at this point because it thought it was the first to start up and is just waiting for others to connect to it
5) Servers 1 and 2 marked server 0 as down and correctly failed over and are both in agreement about the new master (Server 1)
6) Our Rails pods never learn about the failover because they are configured to ask the Sentinel on Server 0 who the master is first; since Sentinel on Server 0 keeps replying successfully that it is the master, the Rails pods never bother to ask Sentinel 1 or Sentinel 2
7) Rails tries to write to Server 0
8) Writes fail because Server 0 has no slaves to replicate to and rejects the write
9) System was stuck in this state indefinitely until we manually rebooted Server 0 and it initialized itself correctly as a slave

**What you expected to happen**:
I expected Server 0 not to believe that it is the master indefinitely when it isn't connected to any of the other nodes. I expected Sentinel on Server 0 to report that it doesn't know who the master is and to eventually report the correct master.

**How to reproduce it** (as minimally and precisely as possible):
1. Cut off network connectivity from the node that hosts the server-0 pod to the nodes hosting server-1 and server-2.
1. Restart the server-0 pod.
1. Restore network connectivity

server-1 or server-2 will be the new master but the sentinel on server-0 is reporting that server-0 is the master and that it is up.

**Anything else we need to know**:

Two possible solutions off the top of my head:

1)  Have a helm config value to indicate a new cluster vs. an existing cluster so that a redis-ha statefulset pod rebooting in an existing cluster will never initialize itself as the master

2) Have a liveness check that eventually kills (and restarts) a statefulset pod that doesn't know about {quorum - 1}  other pods.

### LINKED ISSUE #9568: [stable/redis-ha] implement fullnameOverride option into _helpers.tpl
**Is this a request for help?**: yes

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

Please add upstream the option to specify `fullnameOverride` , to override the fullname of the deploy as is already possibile in many other charts.

example add the following code into _helpers.tpl:

```
{{- define "redis-ha.fullname" -}}
{{- if .Values.fullnameOverride -}}
{{- .Values.fullnameOverride | trunc 63 | trimSuffix "-" -}}
{{- else -}}
{{- $name := default .Chart.Name .Values.nameOverride -}}
{{- if contains $name .Release.Name -}}
{{- .Release.Name | trunc 63 | trimSuffix "-" -}}
{{- else -}}
{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" -}}
{{- end -}}
{{- end -}}
{{- end -}}
```

**Version of Helm and Kubernetes**: Helm 2.11, Kubernetes 1.10

**Which chart**: redis-ha

Kind regards

### PR DESCRIPTION
**Pull request by ssalaues**
#### What this PR does / why we need it:
Major:
Fixes issues where network partitions can result in a split-brain #9488 
  - Better init script that better handles recovery from the split-brain
  - Retained Sentinel IDs
  - Better probes

Minor:
Fixes name override options #9568

---
Part of the root issue of #9488 in that whenever a pod is deleted/recreated, it gets assigned a new IP address causing the remaining quorum to lose track of the down pod after rescheduling. In ideal scenarios this is okay because the init container will find and attach to an existing quorum. However in the case of a network partition this can easily result in the split-brain issue.

These changes leverage the stable IP of a Kubernetes service mapped to the notion of an "Announce IP" in redis/sentinel that allow for more stable operation and discovery. This is necessary because both Redis and Sentinel will resolve DNS names and only persist the corresponding IP in the config not the DNS name. 

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #9568, fixes #9488

#### Special notes for your reviewer:
In my test lab, these fixes allow for the the init to fail and crash, preventing the issue from occurring.

cc @aianus 

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

File: stable/redis-ha/templates/NOTES.txt [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
 {{ template "redis-ha.fullname" . }}.{{ .Release.Namespace }}.svc.cluster.local
 
 To connect to your Redis server:
 
 {{- if .Values.auth }}
-1. Get the randomly generated redis password:
+1. To retrieve the redis password:
+   On MacOS:
    echo $(kubectl get secret {{ template "redis-ha.fullname" . }} -o "jsonpath={.data['auth']}" | base64 -D)
+   On Linux:
```
Comment (giacope) [inline stable/redis-ha/templates/NOTES.txt]: Instead of making a distinction between OSes, you can use `base64 --decode`

**Review by giacope — CHANGES_REQUESTED** [review]

File: stable/redis-ha/templates/redis-ha-statefulset.yaml [hunk 2]
```diff
@@ -47,8 +48,8 @@ spec:
         - sh
         args:
         - /readonly-config/init.sh
-        env:
 {{- if .Values.auth }}
+        env:
```
Comment (kimxogus) [inline stable/redis-ha/templates/redis-ha-statefulset.yaml:52]: this will make `env:` removed when `.Values.auth` is `false`.
```yaml
        args:
        - "/readonly-config/init.sh"
        - name: POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
```
This line should go back to L50.

Comment (ssalaues) [reply stable/redis-ha/templates/redis-ha-statefulset.yaml:52]: Ahh good find! I believe my intention was to delete the POD_IP var as it is no longer necessary. Thank you I will fix this

File: stable/redis-ha/templates/redis-ha-healthchecks.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    app: {{ template "redis-ha.fullname" . }}
+data:
+  check-quorum.sh: |
+    #!/bin/sh
+    set -eu
+    MASTER_GROUP="{{ .Values.redis.masterGroupName }}"
+    SENTINEL_PORT={{ .Values.sentinel.port }}
+    REDIS_PORT={{ .Values.redis.port }}
+    NUM_SLAVES=$(redis-cli -p "$SENTINEL_PORT" sentinel master {{ .Values.redis.masterGroupName }} | awk '/num-slaves/{getline; print}')
+    MIN_SLAVES=$(({{ .Values.sentinel.quorum }} / 2))
```
Comment (aianus) [inline stable/redis-ha/templates/redis-ha-healthchecks.yaml]: I think this should be `{ .Values.redis.config.min-slaves-to-write }`, no?

File: stable/redis-ha/templates/redis-ha-statefulset.yaml [hunk 4]
```diff
@@ -91,21 +96,31 @@ spec:
...
+          valueFrom:
+            secretKeyRef:
+              name: {{ template "redis-ha.fullname" . }}
+              key: auth
+{{- end }}
         livenessProbe:
           exec:
-            command: ["redis-cli", "-p", "{{ .Values.sentinel.port }}", "ping"]
+            command: [ "sh", "/probes/check-quorum.sh", "{{ .Values.sentinel.port }}"]
           initialDelaySeconds: 15
```
Comment (aianus) [inline stable/redis-ha/templates/redis-ha-statefulset.yaml:119]: It will take at least 60 seconds (15s readiness + 3 * 15s liveness failures) for a pod incorrectly initialized as a master to realize it and restart. Seems like a long time.

Could we not check for quorum/slave count in the readiness probe as well to reduce this time to 15 seconds?

**Review by aianus — COMMENTED** [review]
I like how the announce IP addresses are stable now and I think the readiness and liveness checks should make it so the system will automatically recover from a network partition like the one I experienced. I'm a little concerned about how much downtime will still occur in this scenario (at least 60 seconds afaict).

Unfortunately, I'm going on holiday very soon and I don't think I have time to test these changes so my review is based solely on reading the code. I will test it when I'm back (whether it's merged by then or not).

File: stable/redis-ha/templates/redis-ha-statefulset.yaml (hunk 4 above)
Comment (ssalaues) [reply stable/redis-ha/templates/redis-ha-statefulset.yaml:119]: Yeah I agree it could still take too long. I'll change it and test it out. 

Regardless, I have another commit to expose all the liveness/readiness settings in the values.yaml but I left it out of this PR as it made it drastically larger because of the corresponding README changes. I'll open up another PR after this one just for those changes.

File: stable/redis-ha/templates/redis-ha-healthchecks.yaml (hunk 3 above)
Comment (ssalaues) [reply stable/redis-ha/templates/redis-ha-healthchecks.yaml]: That is a good point, I didn't consider that. I'll switch it and add a comment that it is a mandatory field for those that use the `customConfig` that overrides all the `.Values.redis.config` options.

**Comment by ssalaues** [comment]
No worries, thanks for the review @aianus !

**Comment by burnettk** [comment]
@ssalaues @aianus i just noticed this PR and the corresponding issue. we also filed https://github.com/helm/charts/pull/10178, which i think may address the issue. we took a very different approach. curious to see what you all think. we referenced an acceptance test script in the PR that you may find useful. ultimately, we're sad that we have 10 to 20-second "downtimes" during rolling deployments (as can be seen with the acceptance test script) with the current chart, but really happy that some of the edge cases that cause much worse effects are being driven out. thanks!

**Comment by ssalaues** [comment]
@burnettk I was going over that PR and so far I think the PRs can compliment each other for the most part. I like the graceful termination features in particular as this is something I have been worried about in production.

TLDR; I don't think the two are mutually exclusive as the primary take away from this one is to give each pod a stable IP that can be reliably referenced on restart/delete/etc/

File: stable/redis-ha/templates/redis-ha-announce-service.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+{{- $fullName := include "redis-ha.fullname" . }}
+{{- $replicas := int .Values.replicas }}
+{{- $root := . }}
+{{- range $i := until $replicas }}
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ $fullName }}-announce-{{ $i }}
```
Comment (ceshihao) [inline stable/redis-ha/templates/redis-ha-announce-service.yaml:9]: I think maybe these services are not needed, because `StatefulSet` can have `Stable Network ID` through `Headless Service`.
https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id

Comment (ssalaues) [reply stable/redis-ha/templates/redis-ha-announce-service.yaml:9]: @ceshihao This is how the chart is currently using for discovery however the problem comes when a pod gets recreated and is assigned a new IP address (but same name). This causes issues because Redis itself resolves the name and updates the config with only the IP. This causes issues when a network partition occurs and the pod is isolated. With having a service for each pod, it provides a stable IP that the Quorum can always reference and recover missing redis/sentinel even after network partitions. 

These really are only necessary for stable inter pod discovery and all clients should still use the `Headless Service`

**Comment by ssalaues** [comment]
Thank you @unguiculus !
